File: inventory-service/src/main/java/com/ecommerce/inventory/service/InventoryService.java
package com.ecommerce.inventory.service;

import com.ecommerce.inventory.dto.*;
import com.ecommerce.inventory.model.Product;
import com.ecommerce.inventory.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class InventoryService {
    
    private final ProductRepository productRepository;
    
    @Transactional
    public ProductResponse createProduct(ProductRequest request) {
        log.info("Creating new product: {}", request.getName());
        
        if (productRepository.existsBySku(request.getSku())) {
            throw new RuntimeException("Product with SKU already exists: " + request.getSku());
        }
        
        Product product = new Product();
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setSku(request.getSku());
        product.setPrice(request.getPrice());
        product.setStockQuantity(request.getStockQuantity());
        product.setCategory(request.getCategory());
        product.setImageUrl(request.getImageUrl());
        
        Product savedProduct = productRepository.save(product);
        log.info("Product created successfully: {}", savedProduct.getId());
        
        return mapToResponse(savedProduct);
    }
    
    public ProductResponse getProductById(Long id) {
        log.info("Fetching product by ID: {}", id);
        
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found with ID: " + id));
        
        return mapToResponse(product);
    }
    
    public List<ProductResponse> getAllProducts() {
        log.info("Fetching all products");
        
        return productRepository.findAll().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }
    
    public List<ProductResponse> getProductsByCategory(String category) {
        log.info("Fetching products by category: {}", category);
        
        return productRepository.findByCategory(category).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }
    
    @Transactional
    public ProductResponse updateProduct(Long id, ProductRequest request) {
        log.info("Updating product: {}", id);
        
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found with ID: " + id));
        
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setPrice(request.getPrice());
        product.setStockQuantity(request.getStockQuantity());
        product.setCategory(request.getCategory());
        product.setImageUrl(request.getImageUrl());
        
        Product updatedProduct = productRepository.save(product);
        log.info("Product updated successfully: {}", updatedProduct.getId());
        
        return mapToResponse(updatedProduct);
    }
    
    @Transactional
    public void deleteProduct(Long id) {
        log.info("Deleting product: {}", id);
        
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found with ID: " + id));
        
        productRepository.delete(product);
        log.info("Product deleted successfully: {}", id);
    }
    
    // Inter-service communication endpoint
    @Transactional
    public StockCheckResponse checkStock(StockCheckRequest request) {
        log.info("Checking stock for product ID: {} with quantity: {}", 
                request.getProductId(), request.getQuantity());
        
        Product product = productRepository.findById(request.getProductId())
                .orElseThrow(() -> new RuntimeException("Product not found with ID: " + request.getProductId()));
        
        Integer availableQuantity = product.getAvailableQuantity();
        Boolean isAvailable = availableQuantity >= request.getQuantity();
        
        String message;
        if (isAvailable) {
            message = String.format("Stock available. Requested: %d, Available: %d", 
                    request.getQuantity(), availableQuantity);
        } else {
            message = String.format("Insufficient stock. Requested: %d, Available: %d", 
                    request.getQuantity(), availableQuantity);
        }
        
        log.info("Stock check result for product {}: {}", request.getProductId(), message);
        
        return StockCheckResponse.builder()
                .productId(product.getId())
                .available(isAvailable)
                .requestedQuantity(request.getQuantity())
                .availableQuantity(availableQuantity)
                .message(message)
                .build();
    }
    
    @Transactional
    public void reserveStock(Long productId, Integer quantity) {
        log.info("Reserving stock for product ID: {} with quantity: {}", productId, quantity);
        
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with ID: " + productId));
        
        if (product.getAvailableQuantity() < quantity) {
            throw new RuntimeException("Insufficient stock to reserve");
        }
        
        product.setReservedQuantity(product.getReservedQuantity() + quantity);
        productRepository.save(product);
        
        log.info("Stock reserved successfully for product: {}", productId);
    }
    
    @Transactional
    public void releaseStock(Long productId, Integer quantity) {
        log.info("Releasing stock for product ID: {} with quantity: {}", productId, quantity);
        
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with ID: " + productId));
        
        product.setReservedQuantity(Math.max(0, product.getReservedQuantity() - quantity));
        productRepository.save(product);
        
        log.info("Stock released successfully for product: {}", productId);
    }
    
    @Transactional
    public void confirmStock(Long productId, Integer quantity) {
        log.info("Confirming stock for product ID: {} with quantity: {}", productId, quantity);
        
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found with ID: " + productId));
        
        product.setStockQuantity(product.getStockQuantity() - quantity);
        product.setReservedQuantity(Math.max(0, product.getReservedQuantity() - quantity));
        productRepository.save(product);
        
        log.info("Stock confirmed successfully for product: {}", productId);
    }
    
    private ProductResponse mapToResponse(Product product) {
        return ProductResponse.builder()
                .id(product.getId())
                .name(product.getName())
                .description(product.getDescription())
                .sku(product.getSku())
                .price(product.getPrice())
                .stockQuantity(product.getStockQuantity())
                .reservedQuantity(product.getReservedQuantity())
                .availableQuantity(product.getAvailableQuantity())
                .category(product.getCategory())
                .imageUrl(product.getImageUrl())
                .active(product.getActive())
                .createdAt(product.getCreatedAt())
                .build();
    }
}

File: backend/inventory-service/src/services/inventoryService.js
const Product = require('../models/Product');
const { cloudinary } = require('../config/cloudinary');

class InventoryService {
  // Create product
  async createProduct(productData) {
    try {
      const existingProduct = await Product.findOne({ sku: productData.sku });
      if (existingProduct) {
        throw new Error('Product with this SKU already exists');
      }

      const product = new Product(productData);
      await product.save();
      return product;
    } catch (error) {
      throw error;
    }
  }

  // Get product by ID
  async getProductById(productId) {
    try {
      const product = await Product.findById(productId);
      if (!product) {
        throw new Error('Product not found');
      }
      return product;
    } catch (error) {
      throw error;
    }
  }

  // Get all products
  async getAllProducts() {
    try {
      return await Product.find().sort({ createdAt: -1 });
    } catch (error) {
      throw error;
    }
  }

  // Get products by category
  async getProductsByCategory(category) {
    try {
      return await Product.find({ category }).sort({ createdAt: -1 });
    } catch (error) {
      throw error;
    }
  }

  // Update product
  async updateProduct(productId, updateData) {
    try {
      const product = await Product.findByIdAndUpdate(
        productId,
        { ...updateData, updatedAt: new Date() },
        { new: true, runValidators: true }
      );

      if (!product) {
        throw new Error('Product not found');
      }

      return product;
    } catch (error) {
      throw error;
    }
  }

  // Delete product
  async deleteProduct(productId) {
    try {
      const product = await Product.findByIdAndDelete(productId);
      if (!product) {
        throw new Error('Product not found');
      }
      return { message: 'Product deleted successfully' };
    } catch (error) {
      throw error;
    }
  }

  // Check stock availability (for inter-service communication)
  async checkStock(productId, quantity) {
    try {
      const product = await Product.findById(productId);
      
      if (!product) {
        return {
          available: false,
          message: 'Product not found',
          productId
        };
      }

      if (!product.available) {
        return {
          available: false,
          message: 'Product is not available',
          productId,
          productName: product.name
        };
      }

      const availableQuantity = product.quantity - product.reservedQuantity;
      
      if (availableQuantity < quantity) {
        return {
          available: false,
          message: 'Insufficient stock',
          productId,
          productName: product.name,
          requestedQuantity: quantity,
          availableQuantity
        };
      }

      return {
        available: true,
        message: 'Stock available',
        productId,
        productName: product.name,
        price: product.price,
        requestedQuantity: quantity,
        availableQuantity
      };
    } catch (error) {
      throw error;
    }
  }

  // Reserve stock (for order processing)
  async reserveStock(productId, quantity) {
    try {
      const product = await Product.findById(productId);
      
      if (!product) {
        throw new Error('Product not found');
      }

      const availableQuantity = product.quantity - product.reservedQuantity;
      
      if (availableQuantity < quantity) {
        throw new Error('Insufficient stock to reserve');
      }

      product.reservedQuantity += quantity;
      await product.save();

      return {
        success: true,
        message: 'Stock reserved successfully',
        productId,
        reservedQuantity: quantity,
        totalReserved: product.reservedQuantity
      };
    } catch (error) {
      throw error;
    }
  }

  // Confirm stock (deduct from inventory after payment)
  async confirmStock(productId, quantity) {
    try {
      const product = await Product.findById(productId);
      
      if (!product) {
        throw new Error('Product not found');
      }

      if (product.reservedQuantity < quantity) {
        throw new Error('Reserved quantity mismatch');
      }

      product.quantity -= quantity;
      product.reservedQuantity -= quantity;
      await product.save();

      return {
        success: true,
        message: 'Stock confirmed and deducted',
        productId,
        deductedQuantity: quantity,
        remainingQuantity: product.quantity
      };
    } catch (error) {
      throw error;
    }
  }

  // Release reserved stock (if order fails)
  async releaseStock(productId, quantity) {
    try {
      const product = await Product.findById(productId);
      
      if (!product) {
        throw new Error('Product not found');
      }

      product.reservedQuantity = Math.max(0, product.reservedQuantity - quantity);
      await product.save();

      return {
        success: true,
        message: 'Reserved stock released',
        productId,
        releasedQuantity: quantity
      };
    } catch (error) {
      throw error;
    }
  }

  // Upload product image to Cloudinary
  async uploadProductImage(productId, file) {
    try {
      const product = await Product.findById(productId);
      
      if (!product) {
        throw new Error('Product not found');
      }

      // Delete old image if exists
      if (product.imagePublicId) {
        await cloudinary.uploader.destroy(product.imagePublicId);
      }

      // Upload new image
      const result = await new Promise((resolve, reject) => {
        const uploadStream = cloudinary.uploader.upload_stream(
          {
            folder: 'ecommerce-products',
            transformation: [
              { width: 800, height: 800, crop: 'limit' },
              { quality: 'auto' }
            ]
          },
          (error, result) => {
            if (error) reject(error);
            else resolve(result);
          }
        );
        uploadStream.end(file.buffer);
      });

      // Update product with new image URL
      product.imageUrl = result.secure_url;
      product.imagePublicId = result.public_id;
      await product.save();

      return {
        success: true,
        message: 'Image uploaded successfully',
        imageUrl: product.imageUrl,
        product
      };
    } catch (error) {
      throw error;
    }
  }

  // Delete product image from Cloudinary
  async deleteProductImage(productId) {
    try {
      const product = await Product.findById(productId);
      
      if (!product) {
        throw new Error('Product not found');
      }

      if (!product.imagePublicId) {
        throw new Error('Product has no image to delete');
      }

      // Delete from Cloudinary
      await cloudinary.uploader.destroy(product.imagePublicId);

      // Remove from product
      product.imageUrl = null;
      product.imagePublicId = null;
      await product.save();

      return {
        success: true,
        message: 'Image deleted successfully',
        product
      };
    } catch (error) {
      throw error;
    }
  }
}

module.exports = new InventoryService();
